dindex_store/discovery_index.py

```diff
@@ -0,0 +1,178 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Dict, List
+
+
+class EdgeType(Enum):
+    ATTRIBUTE_SYNTACTIC_SIMILARITY = 0
+    COLUMN_CONTENT_JACCARD_SIMILARITY = 1
+    SYNTACTIC_INCLUSION_DEPENDENCY = 2
+
+
+class DiscoveryIndex:
+    """
+    DiscoveryIndex stores profiles generated by the profiler, a graph that
+    describes relationships between profiles, and a full text search index on
+    the original input data.
+
+    :param profile_index: ProfileIndex
+    :param graph_idnex: GraphIndex
+    :param fts_index: FullTextSearchIndex
+    """
+
+    def __init__(self, profile_index, graph_index, fts_index=None) -> None:
+        self.__profile_index = profile_index
+        self.__graph_index = graph_index
+        self.__fts_index = fts_index
+
+    # ----------------------------------------------------------------------
+    # Modify Methods
+
+    def add_node(self, node: Dict) -> bool:
+        node_id = node["id"]
+        if not self.__profile_index.add_profile(node):
+            return False
+        return self.__graph_index.add_node(node_id)
+
+    def add_edge(
+            self,
+            source_node_id: int,
+            target_node_id: int,
+            type: EdgeType,
+            properties: Dict) -> bool:
+        return self.__graph_index.add_edge(
+            source_node_id, target_node_id, type, properties)
+
+    def add_undirected_edge(
+            self,
+            source_node_id: int,
+            target_node_id: int,
+            type: EdgeType,
+            properties: Dict) -> bool:
+        return self.__graph_index.add_undirected_edge(
+            source_node_id, target_node_id, type, properties)
+
+    # ----------------------------------------------------------------------
+    # Query Methods
+
+    def get_node(self, node_id: int) -> Dict:
+        return self.__profile_index.get_node(node_id)
+
+    def find_neighborhood(self, node_id: int, hops: int = 1):
+        return self.__graph_index.find_neighborhood(node_id, hops)
+
+    def find_path(
+            self,
+            source_node_id: int,
+            target_node_id: int,
+            max_len: int = 3):
+        return self.__graph_index.find_neighborhood(
+            source_node_id, target_node_id, max_len)
+
+
+class ProfileIndex(ABC):
```

Comment:
I like this design where DiscoveryIndex is concrete, and the other, underlying indices are abstract.